[MSVC][ARM64EC] Failed to build with error: cannot convert argument 1 from 'const DirectX::XMVECTOR' to '__m128'

DirectXMath is encountering the following error when building the project with MSVC in arm64ec architecture:
```c++
error C2664: '__m128 _mm_fmadd_ps(__m128,__m128,__m128)': 
cannot convert argument 1 from 'const DirectX::XMVECTOR' to '__m128' 
```
This issue occurs to the header files in the [Extension/](https://github.com/microsoft/DirectXMath/tree/main/Extensions).

Operating System: Windows

Step to reproduce:

1. Open X64 Native Tools Command Prompt
2. git clone https://github.com/microsoft/DirectXMath.git
3. cd DirectXMath
4. git clone https://github.com/walbourn/directxmathtest.git Tests
6.  set VSCMD_SKIP_SENDTELEMETRY=1 & "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\VsDevCmd.bat" -host_arch=amd64 -arch=arm64
7. cl /c /IC:\gitP\Microsoft\DirectXMath\Tests\..\Inc /IC:\gitP\Microsoft\DirectXMath\Tests\..\Extensions /nologo /Wall /WX- /diagnostics:column /O2 /Ob2 /Oy- /D _AMD64_ /D _ARM64EC_ /D _UNICODE /D UNICODE /D WIN32 /D _AMD64_ /D AMD64 /D _ARM64EC_ /D ARM64EC /D _WINDOWS /D NDEBUG /D _UNICODE /D UNICODE /D "CMAKE_INTDIR=\"Release\"" /Gm- /EHsc /MT /GS /arch:AVX2 /fp:fast /Zc:wchar_t /Zc:forScope /Zc:inline /GR- /std:c++17 /permissive-  /external:W4 /Gd /TP /wd4061 /wd4365 /wd4514 /wd4710 /wd4820 /wd4668 /wd5039 /wd5045 /wd5264 /analyze- /errorReport:queue /arm64EC  /Zc:__cplusplus C:\gitP\Microsoft\DirectXMath\Tests\ext\testfma3.cpp

or simply run this:

1. Download the preprocessed file [testfma3.txt](https://github.com/user-attachments/files/16279141/testfma3.txt)
2. Open X64 Native Tools Command Prompt
3. set VSCMD_SKIP_SENDTELEMETRY=1 & "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\VsDevCmd.bat" -host_arch=amd64 -arch=arm64
4. cl /c /nologo  /D _ARM64EC_ /Gm- /EHsc /std:c++17 /permissive- /TP testfma3.txt



```[tasklist]
### Tasks
```

--- a/CMakePresets.json
+++ b/CMakePresets.json
@@ -47,6 +47,15 @@
       "cacheVariables": { "DXMATH_ARCHITECTURE": "arm64" },
       "hidden": true
     },
+    {
+      "name": "ARM64EC",
+      "architecture": {
+        "value": "arm64ec",
+        "strategy": "external"
+      },
+      "cacheVariables": { "DXMATH_ARCHITECTURE": "arm64ec" },
+      "hidden": true
+    },
 
     {
       "name": "Debug",
@@ -149,6 +158,8 @@
     { "name": "arm-Release"  , "description": "MSVC for ARM (Release) - ARM-NEON", "inherits": [ "base", "ARM", "Release", "MSVC" ] },
     { "name": "arm64-Debug"  , "description": "MSVC for ARM64 (Debug) - ARM-NEON", "inherits": [ "base", "ARM64", "Debug", "MSVC" ] },
     { "name": "arm64-Release", "description": "MSVC for ARM64 (Release) - ARM-NEON", "inherits": [ "base", "ARM64", "Release", "MSVC" ] },
+    { "name": "arm64ec-Debug"  , "description": "MSVC for ARM64EC (Debug) - ARM-NEON", "inherits": [ "base", "ARM64EC", "Debug", "MSVC" ], "environment": { "CXXFLAGS": "/arm64EC" } },
+    { "name": "arm64ec-Release", "description": "MSVC for ARM64EC (Release) - ARM-NEON", "inherits": [ "base", "ARM64EC", "Release", "MSVC" ], "environment": { "CXXFLAGS": "/arm64EC" } },
 
     { "name": "x64-Debug-Clang"    , "description": "Clang/LLVM for x64 (Debug) - SSE/SSE2", "inherits": [ "base", "x64", "Debug", "Clang" ] },
     { "name": "x64-Release-Clang"  , "description": "Clang/LLVM for x64 (Release) - SSE/SSE2", "inherits": [ "base", "x64", "Release", "Clang" ] },

--- a/Extensions/DirectXMathAVX.h
+++ b/Extensions/DirectXMathAVX.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) || __arm__ || __aarch64__
 #error AVX not supported on ARM platform
 #endif
 

--- a/Extensions/DirectXMathAVX2.h
+++ b/Extensions/DirectXMathAVX2.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) || __arm__ || __aarch64__
 #error AVX2 not supported on ARM platform
 #endif
 

--- a/Extensions/DirectXMathBE.h
+++ b/Extensions/DirectXMathBE.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if (defined(_M_IX86) || defined(_M_X64) || __i386__ || __x86_64__) && !defined(_M_HYBRID_X86_ARM64)
+#if (defined(_M_IX86) || defined(_M_X64) || __i386__ || __x86_64__) && !defined(_M_HYBRID_X86_ARM64) && !defined(_M_ARM64EC)
 #include <tmmintrin.h>
 #endif
 

--- a/Extensions/DirectXMathF16C.h
+++ b/Extensions/DirectXMathF16C.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) || __arm__ || __aarch64__
 #error F16C not supported on ARM platform
 #endif
 

--- a/Extensions/DirectXMathFMA3.h
+++ b/Extensions/DirectXMathFMA3.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) || __arm__ || __aarch64__
 #error FMA3 not supported on ARM platform
 #endif
 

--- a/Extensions/DirectXMathFMA4.h
+++ b/Extensions/DirectXMathFMA4.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) || __arm__ || __aarch64__
 #error FMA4 not supported on ARM platform
 #endif
 

--- a/Extensions/DirectXMathSSE3.h
+++ b/Extensions/DirectXMathSSE3.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) || __arm__ || __aarch64__
 #error SSE3 not supported on ARM platform
 #endif
 

--- a/Extensions/DirectXMathSSE4.h
+++ b/Extensions/DirectXMathSSE4.h
@@ -9,7 +9,7 @@
 
 #pragma once
 
-#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || __arm__ || __aarch64__
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_HYBRID_X86_ARM64) || defined(_M_ARM64EC) || __arm__ || __aarch64__
 #error SSE4 not supported on ARM platform
 #endif
 

--- a/build/DirectXMath-GitHub-Dev17.yml
+++ b/build/DirectXMath-GitHub-Dev17.yml
@@ -345,3 +345,151 @@ jobs:
       platform: x64
       configuration: Release
       msbuildArchitecture: x64
+
+- job: CMAKE_BUILD
+  displayName: 'CMake BUILD_TESTING=ON'
+  timeoutInMinutes: 60
+  workspace:
+    clean: all
+  steps:
+  - checkout: self
+    clean: true
+    fetchTags: false
+  - task: DeleteFiles@1
+    displayName: Delete files from Tests
+    inputs:
+      SourceFolder: Tests
+      Contents: '**'
+      RemoveSourceFolder: true
+      RemoveDotFiles: true
+  - task: CmdLine@2
+    displayName: Fetch Tests
+    inputs:
+      script: git clone --quiet --no-tags https://%GITHUB_PAT%@github.com/walbourn/directxtktest.git Tests
+      workingDirectory: $(Build.SourcesDirectory)
+      failOnStderr: true
+  - task: CmdLine@2
+    displayName: Setup environment for CMake to use VS
+    inputs:
+      script: |
+        call "$(VC_PATH)\Auxiliary\Build\vcvars64.bat"
+        echo ##vso[task.setvariable variable=WindowsSdkVerBinPath;]%WindowsSdkVerBinPath%
+        echo ##vso[task.prependpath]%VSINSTALLDIR%Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja
+        echo ##vso[task.prependpath]%VCINSTALLDIR%Tools\Llvm\x64\bin
+        echo ##vso[task.prependpath]%WindowsSdkBinPath%x64
+        echo ##vso[task.prependpath]%WindowsSdkVerBinPath%x64
+        echo ##vso[task.prependpath]%VCToolsInstallDir%bin\Hostx64\x64
+        echo ##vso[task.setvariable variable=EXTERNAL_INCLUDE;]%EXTERNAL_INCLUDE%
+        echo ##vso[task.setvariable variable=INCLUDE;]%INCLUDE%
+        echo ##vso[task.setvariable variable=LIB;]%LIB%
+
+  - task: CMake@1
+    displayName: CMake (MSVC; x64-Debug) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=x64-Debug
+  - task: CMake@1
+    displayName: CMake (MSVC; x64-Debug) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/x64-Debug -v
+  - task: CMake@1
+    displayName: CMake (MSVC; x64-Release) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=x64-Release
+  - task: CMake@1
+    displayName: CMake (MSVC; x64-Release) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/x64-Release -v
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; x64-Debug) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=x64-Debug-Clang
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; x64-Debug) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/x64-Debug-Clang -v
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; x64-Release) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=x64-Release-Clang
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; x64-Release) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/x64-Release-Clang -v
+  - task: DeleteFiles@1
+    inputs:
+      Contents: 'out'
+  - task: CmdLine@2
+    displayName: Switch compiler to ARM64
+    inputs:
+      script: |
+        call "$(VC_PATH)\Auxiliary\Build\vcvarsamd64_arm64.bat"
+        echo ##vso[task.setvariable variable=LIB;]%LIB%
+
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64-Debug) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=arm64-Debug
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64-Debug) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/arm64-Debug -v
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64-Release) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=arm64-Release
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64-Release) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/arm64-Release -v
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; arm64-Debug) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=arm64-Debug-Clang
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; arm64-Debug) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/arm64-Debug-Clang -v
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; arm64-Release) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=arm64-Release-Clang
+  - task: CMake@1
+    displayName: CMake (clang/LLVM; arm64-Release) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/arm64-Release-Clang -v
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64ec-Debug) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=arm64ec-Debug
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64ec-Debug) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/arm64ec-Debug -v
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64ec-Release) Config
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --preset=arm64ec-Release
+  - task: CMake@1
+    displayName: CMake (MSVC; arm64ec-Release) Build
+    inputs:
+      cwd: '$(Build.SourcesDirectory)'
+      cmakeArgs: --build out/build/arm64-Release -v

--- a/build/DirectXMath-GitHub-Dev17.yml
+++ b/build/DirectXMath-GitHub-Dev17.yml
@@ -454,6 +454,7 @@ jobs:
         call "$(VC_PATH)\Auxiliary\Build\vcvarsamd64_arm64.bat"
         echo ##vso[task.setvariable variable=WindowsSdkVerBinPath;]%WindowsSdkVerBinPath%
         echo ##vso[task.prependpath]%VSINSTALLDIR%Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja
+        echo ##vso[task.prependpath]%VCToolsInstallDir%\bin\Hostx64\arm64
         echo ##vso[task.prependpath]%VCINSTALLDIR%Tools\Llvm\x64\bin
         echo ##vso[task.prependpath]%WindowsSdkBinPath%x64
         echo ##vso[task.prependpath]%WindowsSdkVerBinPath%x64

--- a/build/DirectXMath-GitHub-Dev17.yml
+++ b/build/DirectXMath-GitHub-Dev17.yml
@@ -521,4 +521,4 @@ jobs:
     displayName: CMake (MSVC; arm64ec-Release) Build
     inputs:
       cwd: '$(Build.SourcesDirectory)'
-      cmakeArgs: --build out/build/arm64-Release -v
+      cmakeArgs: --build out/build/arm64ec-Release -v

--- a/build/DirectXMath-GitHub-Dev17.yml
+++ b/build/DirectXMath-GitHub-Dev17.yml
@@ -52,6 +52,7 @@ pool:
 
 variables:
   Codeql.Enabled: false
+  VC_PATH: 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC'
   GITHUB_PAT: $(GITHUBPUBLICTOKEN)
 
 jobs:

--- a/build/DirectXMath-GitHub-Dev17.yml
+++ b/build/DirectXMath-GitHub-Dev17.yml
@@ -347,8 +347,8 @@ jobs:
       configuration: Release
       msbuildArchitecture: x64
 
-- job: CMAKE_BUILD
-  displayName: 'CMake BUILD_TESTING=ON'
+- job: CMAKE_BUILD_X64
+  displayName: 'CMake for X64 BUILD_TESTING=ON'
   timeoutInMinutes: 60
   workspace:
     clean: all
@@ -424,14 +424,42 @@ jobs:
     inputs:
       cwd: '$(Build.SourcesDirectory)'
       cmakeArgs: --build out/build/x64-Release-Clang -v
+
+- job: CMAKE_BUILD_ARM64
+  displayName: 'CMake for ARM64 BUILD_TESTING=ON'
+  timeoutInMinutes: 60
+  workspace:
+    clean: all
+  steps:
+  - checkout: self
+    clean: true
+    fetchTags: false
   - task: DeleteFiles@1
+    displayName: Delete files from Tests
     inputs:
-      Contents: 'out'
+      SourceFolder: Tests
+      Contents: '**'
+      RemoveSourceFolder: true
+      RemoveDotFiles: true
   - task: CmdLine@2
-    displayName: Switch compiler to ARM64
+    displayName: Fetch Tests
+    inputs:
+      script: git clone --quiet --no-tags https://%GITHUB_PAT%@github.com/walbourn/directxmathtest.git Tests
+      workingDirectory: $(Build.SourcesDirectory)
+      failOnStderr: true
+  - task: CmdLine@2
+    displayName: Setup environment for CMake to use VS
     inputs:
       script: |
         call "$(VC_PATH)\Auxiliary\Build\vcvarsamd64_arm64.bat"
+        echo ##vso[task.setvariable variable=WindowsSdkVerBinPath;]%WindowsSdkVerBinPath%
+        echo ##vso[task.prependpath]%VSINSTALLDIR%Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja
+        echo ##vso[task.prependpath]%VCINSTALLDIR%Tools\Llvm\x64\bin
+        echo ##vso[task.prependpath]%WindowsSdkBinPath%x64
+        echo ##vso[task.prependpath]%WindowsSdkVerBinPath%x64
+        echo ##vso[task.prependpath]%VCToolsInstallDir%bin\Hostx64\x64
+        echo ##vso[task.setvariable variable=EXTERNAL_INCLUDE;]%EXTERNAL_INCLUDE%
+        echo ##vso[task.setvariable variable=INCLUDE;]%INCLUDE%
         echo ##vso[task.setvariable variable=LIB;]%LIB%
 
   - task: CMake@1

--- a/build/DirectXMath-GitHub-Dev17.yml
+++ b/build/DirectXMath-GitHub-Dev17.yml
@@ -454,11 +454,10 @@ jobs:
         call "$(VC_PATH)\Auxiliary\Build\vcvarsamd64_arm64.bat"
         echo ##vso[task.setvariable variable=WindowsSdkVerBinPath;]%WindowsSdkVerBinPath%
         echo ##vso[task.prependpath]%VSINSTALLDIR%Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja
-        echo ##vso[task.prependpath]%VCToolsInstallDir%\bin\Hostx64\arm64
         echo ##vso[task.prependpath]%VCINSTALLDIR%Tools\Llvm\x64\bin
         echo ##vso[task.prependpath]%WindowsSdkBinPath%x64
         echo ##vso[task.prependpath]%WindowsSdkVerBinPath%x64
-        echo ##vso[task.prependpath]%VCToolsInstallDir%bin\Hostx64\x64
+        echo ##vso[task.prependpath]%VCToolsInstallDir%bin\Hostx64\arm64
         echo ##vso[task.setvariable variable=EXTERNAL_INCLUDE;]%EXTERNAL_INCLUDE%
         echo ##vso[task.setvariable variable=INCLUDE;]%INCLUDE%
         echo ##vso[task.setvariable variable=LIB;]%LIB%

--- a/build/DirectXMath-GitHub-Dev17.yml
+++ b/build/DirectXMath-GitHub-Dev17.yml
@@ -366,7 +366,7 @@ jobs:
   - task: CmdLine@2
     displayName: Fetch Tests
     inputs:
-      script: git clone --quiet --no-tags https://%GITHUB_PAT%@github.com/walbourn/directxtktest.git Tests
+      script: git clone --quiet --no-tags https://%GITHUB_PAT%@github.com/walbourn/directxmathtest.git Tests
       workingDirectory: $(Build.SourcesDirectory)
       failOnStderr: true
   - task: CmdLine@2

